### PR TITLE
[Material Design] Side menu design for builds page

### PR DESCRIPTION
**Pull request by shanzi**
Now finally we are working on the most important page: builds page. I come up with a navigation design for it. At first, `masters`, `slaves`, `schedulers` and `changes` along with all builders separately will become sub state of `builds` route. I put a side menu on the page so that we get a split view. The side menu has been almost finished now and I have already configured routes for these sub states.

A preview of the side menu is:

builds_menu_design

Highlighting of `menu-item` of current displaying sub state is done too.

@tardyp @tothandras I haven't switch to new data services. Please review the code and have a look at the [live demo](http://buildbot.apps.io-meter.com) first. Once @tothandras 's latest work has been merged, I will start the migration. :)

### CODE REVIEW

File: www/md_base/src/app/builds/builds.tpl.jade [hunk 1]
```diff
@@ -1 +1,47 @@
...
+      )
+        div(flex)
+          a.builder-title(ui-sref="builds.builder({builderid:{{builder.builderid}}})") {{ builder.name }}
+          div.tags
+            span.tag(ng-repeat="tag in builder.tags") {{tag}}
+        div.builder-id \#{{ builder.builderid }}
+
+md-content.sub-content(flex, ui-view="")
+  div.placeholder
+    | Choose something to view at the panel left.
```
Comment (tothandras) [inline www/md_base/src/app/builds/builds.tpl.jade]: Did you try to use [abstract states](https://github.com/angular-ui/ui-router/wiki/Nested-States-%26-Nested-Views#abstract-states) instead? I would show something by default.

File: www/md_base/src/app/builds/builds.controller.coffee [hunk 2]
```diff
@@ -1,4 +1,11 @@
 class Builds extends Controller
+    builders: []
+
+    isHighlighted: (name, param) ->
+        return @$state.is(name, param)
+
+    constructor: (@$state, buildbotService) ->
+        buildbotService.all('builders').getList().then (builders) =>
+            @builders = builders
```
Comment (tothandras) [inline www/md_base/src/app/builds/builds.controller.coffee]: .then (@builders) =>

File: www/md_base/src/app/builds/builder/builder.route.coffee [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+
+        # Register new state
+        $stateProvider.state
+            controller: "builderController"
+            controllerAs: 'builder'
+            templateUrl: "views/#{name}.html"
+            name: name
+            url: "/builder/:builderid"
+            data:
+                title: 'Builds / Builder'
```
Comment (tothandras) [inline www/md_base/src/app/builds/builder/builder.route.coffee]: The title could be just `Builder` and you can use the parent state to get `Builds / Builder`.

**Comment by tardyp** [comment]
We will need filtering by category. Some installs have 400+ builders. Having the builder list only on the sidebar looks it will be complicated for the users to find the one they want. Also, this sidebar is taking a lot of horizontal space, so at first sight I was a bit unsure. Lets see how the build page will look like.

**Comment by shanzi** [comment]
Added a input for filtering builders. 

filter_input

Now with the help from @tothandras the new UI is using latest `dataService` :)

@tardyp  As for further design of the builds page. I'd like to stick to what I thought of before. But due to some limits of current API service, some changes will be made.

A preview of previous design:

File: www/md_base/src/app/builds/builds.tpl.jade (hunk 1 above)
Comment (shanzi) [inline www/md_base/src/app/builds/builds.tpl.jade]: I think add something directly to the `ui-view` as placeholder is good enough :)

**Comment by shanzi** [comment]
@tardyp @tothandras More comments on this PR?

I'd like to get this PR merged first if there are no other problems left and go on developing on builds page. I plan to finish masters, slaves, schedulers and changes page next and put the builders page at last.

Hope we can finish the former four pages before mid-term evaluation :)

File: www/md_base/src/app/common/directives/builditem/builditem.directive.coffee [hunk 4]
```diff
@@ -14,12 +14,10 @@ class BuildItem extends Directive
...
+    constructor: (dataService) ->
         # TODO: add links to builders and builds after the corresponding pages has finished
         if @outBuilder
             @builder = outBuilder
         else if @showBuilder
-            buildbotService.one('builders', @build.builderid).get().then (data) =>
-                @builder = data
-
-
+            dataService.getBuilders(@build.builderid).then (builders) =>
```
Comment (tothandras) [inline www/md_base/src/app/common/directives/builditem/builditem.directive.coffee]: I messed this up, it was me, right? It won't clean up the listeners. You can pass in subscribe: false to it, and it won't listen for changes.

Comment (shanzi) [inline www/md_base/src/app/common/directives/builditem/builditem.directive.coffee]: `dataService.getBuilders(@build.builderid, subscribe: false)` ?

**Comment by shanzi** [comment]
I added a dropdown menu for narrow viewport:

dropdown_menu

**Comment by tardyp** [comment]
not merging, as the demo server looks broken.

**Comment by tothandras** [comment]
It's something to do with minification.

**Comment by tothandras** [comment]
It's me, something to do with buildbot-data. I am not sure where, I am fixing it. I'll notify you.

**Comment by tothandras** [comment]
I've got it @shanzi, please update to 1.0.8. Sorry for the trouble.

**Comment by shanzi** [comment]
Done!

@tothandras Never mind. I have met a similar problem before. The js uglify sometimes generated code in an unexpected way. :)

**Comment by shanzi** [comment]
Please merge this :)

**Pull request merged**